lib/rubygems/core_ext/kernel_warn.rb

```diff
@@ -36,20 +28,25 @@ class << self
             break
           end
 
-          start += 1
-
           if path = loc.path
             unless path.start_with?(rubygems_path) or path.start_with?('<internal:')
               # Non-rubygems frames
               uplevel -= 1
+              break if uplevel < 0
             end
           end
+
+          start += 1
         end
-        uplevel = start
+        kw[:uplevel] = start
       end
 
-      kw[:uplevel] = uplevel
-      original_warn.call(*messages, **kw)
-    }
+      super(*messages, **kw)
+    end
   end
+
+  Kernel.prepend(KernelExt)
+  Object.include(Kernel)
```

Comment:
This should be unnecessary.

Reply:
Did you try it? I think it's necessary because if `Kernel` has already been included into `Object`, the enhanced `Kernel` won't take effect on `warn` calls without a receiver.

Reply:
I'll try it tomorrow.
Changing `Kernel#warn` is quite tricky, much harder than `Warning.warn`, I wish there would be a proper API for that.

One downside of this approach is it adds an extra entry in ancestors for every `Object`. That slows down every method lookup that would read a Kernel method or above.
And I guess it won't work for things like `Delegator < BasicObject` which `include Kernel` (although Delegator makes a copy of Kernel so that case is more complex than that).

One idea would be to still monkey patch, but to call the original Kernel#warn with `Warning` as self when self is `Warning` to avoid the recursion. In other words, to preserve the value of `self` through this override. It could also be done with an UnboundMethod and then `bind(self).call()`.

Reply:
I see. Would you be able to propose a follow up patch fixing this?

Reply:
I changed it to extend `Object` instead of prepending. Should it better since it does not modify `Object`'s ancestors chain?

Reply:
I just made https://github.com/rubygems/rubygems/pull/3987

Reply:
Could you review that one? I think it's the safest approach with the least side effects.
`prepend` / `extend` look nicer, but they don't achieve the same effect and they have various corner cases where they don't work (for modules like Kernel, for classes it would work better).

Reply:
Thanks!